---
title: "Intro to Transcriptomics"
author: Nicole Black, Wade Boohar, Kayla Xu
date: 07/17/22
updated: 09/26/23
---

***Deliverables***
-Upload this R Notebook to your GitHub and submit the link to your Repo on BB.
-Include any graphs or figures created in this assignment in the folder with your R notebook with descriptive file names.

Since this is an optional partner activity, it is okay if your answers are the same as your partnerâ€™s as long as everyone understands it and could explain it in their own words if asked. Each person must individually push their code to Github. *At the top of your R Notebook, write the name of you and your partner(s) as a comment.*

***Complete the following coding activity and answer any following questions as comments in your R Notebook***

In Transcriptomics_Tutorial.Rmd, you learned how to manipulate the SummarizedExperiment data structure, turn it into more readable dataframes, complete basic analysis using those dataframes, and performed a differential expression analysis based on age. Before saving the rna_counts, rna_clinical, and rna_genes however, we masked out any sample that was not a tumor. In this assignment, you will need to reinstantiate the rna_se object and then perform differential expression analysis based on tumor status.

*Pre-Assignment*
Use knitr function to set your working directory to your analysis_data folder in qbio_490_name.
```{r}
knitr::opts_knit$set(root.dir = normalizePath("/Users/abwilk/Documents/Fall 2023/QBIO 490/qbio_490_andrewwilk/analysis_data")) 
```

Load all necessary packages and instantiate the rna_se object
```{r}
if (!require("TCGAbiolinks", quietly = TRUE))
BiocManager::install("TCGAbiolinks")
if (!require("SummarizedExperiment", quietly = TRUE))
BiocManager::install("SummarizedExperiment")

library(SummarizedExperiment)
library(TCGAbiolinks)

rna_query <- GDCquery(project ="TCGA-BRCA",
                      data.category = "Transcriptome Profiling",
                      data.type = "Gene Expression Quantification",
                      workflow.type = "STAR - Counts")

rna_se <- GDCprepare(rna_query)
```


*1*
Perform all data cleaning and label manipulation steps (hint: applying colnames and rownames to make the dataframes more readable) except for masking out the patient with an NA for age and the "Solid Tissue Normal" sample types.
```{r}
rna_clinical <- rna_se@colData
rna_clinical <- as.data.frame(rna_clinical)

treatments_mask <- ifelse(colnames(rna_clinical) == 'treatments', F,T)
rna_clinical <- rna_clinical[, treatments_mask]
primary_site_mask <- ifelse(colnames(rna_clinical) == 'primary_site', F,T)
rna_clinical <- rna_clinical[, primary_site_mask]
disease_type_mask <- ifelse(colnames(rna_clinical) == 'disease_type', F,T)
rna_clinical <- rna_clinical[, disease_type_mask]

rna_genes <- rna_se@rowRanges@elementMetadata
rna_genes <- as.data.frame(rna_genes)

rna_counts <- rna_se@assays@data$unstranded
rna_counts <- as.data.frame(rna_counts)

rna_clinical$age_category <- ifelse(rna_clinical$age_at_index <=58, 'young','old')

rownames(rna_genes) <- rna_genes$gene_id

rownames(rna_counts) <- rownames(rna_genes)
colnames(rna_counts) <- rownames(rna_clinical)

```


*2*
<<<<<<< HEAD
Determine and define your experimental design for the differential expression analysis. Control for covariance of age, gender, OR vital_status. Manipulate those columns so that they are ready for differential expression analysis (hint: what kind of variables are they? what data type are they by default? do you need to handle unknown values?)
=======
Determine and define your experimental design for the differential expression analysis. Control for covariance of at least 2 other variables. Manipulate those columns so that they are ready for differential expression analysis (hint: what kind of variables are they? what kind of variable do they need to be? is the data type correct? do you need to handle unknown values?)
>>>>>>> 75150ea0bc4dd3d410d7944d4b31b80ff89c3865
```{r}

# Going to do differential expression analysis comparing black vs. white patients. Going to control for covariance for age and vital status.

black_or_white_mask <- ifelse(rna_clinical$race == 'white' | rna_clinical$race == 'black or african american',T,F)
rna_clinical <- rna_clinical[black_or_white_mask,]

na_mask <- ifelse(!is.na(rna_clinical$race) == TRUE & !is.na(rna_clinical$vital_status) == TRUE &       !is.na(rna_clinical$age_category) == TRUE, T, F)
rna_clinical <- rna_clinical[na_mask,]

rna_counts <- rna_counts[,colnames(rna_counts) %in% rownames(rna_clinical)]


rna_clinical$race <- factor(rna_clinical$race)
head(rna_clinical$race)
rna_clinical$age_category <- factor(rna_clinical$age_category)
rna_clinical$vital_status <- factor(rna_clinical$vital_status)

rna_clinical$age_category <- factor(rna_clinical$age_category, levels = c('young','old'))
head(rna_clinical$age_category)

row_sums <- rowSums(rna_counts, na.rm = TRUE)
low_counts_mask <- ifelse(row_sums >= 10, T,F)

rna_counts <- rna_counts[low_counts_mask,]
rna_genes <- rna_genes[low_counts_mask,]

```


*3*
Perform the differential expression analysis, create a volcano plot, and interpret what genes from each of the 6 sections means (assuming thresholds of -log10padj = 2 and log2FoldChange = 1)
```{r}
if (!require("DESeq2", quietly = TRUE))
BiocManager::install("DESeq2")
library(DESeq2)

dds <- DESeqDataSetFromMatrix(countData = rna_counts,
                             colData = rna_clinical,
                             design = ~age_category + vital_status + race)
                             
dds_obj <- DESeq(dds)

resultsNames(dds_obj)

results <- results(dds_obj, format = "DataFrame", contrast = c("race", 'white', 'black or african american'))

# Volcano Plot

if (!require("EnhancedVolcano", quietly = TRUE))
BiocManager::install("EnhancedVolcano")
library(EnhancedVolcano)

EnhancedVolcano(results,
    lab = rownames(results),
    x = 'log2FoldChange',
    y = 'pvalue')


```

top-right genes: significantly upregulated
bottom-right genes: non-significantly upregulated
top-left genes: significantly downregulated
bottom-left genes: non-significantly downregulated
top-middle genes: significant but not differentially expressed
bottom-middle genes: non-significant not differentially expressed

Save the picture of the volcano plot (using either ggsave() or right clicking and manually downloading the image and push this .Rmd and the image to GitHub)
